vicinity/backends/basic.py

```diff
@@ -107,43 +124,98 @@ def query(
+        Batched distance query.
+
+        :param vectors: The vectors to query.
+        :param k: The number of nearest neighbors to return.
+        :return: A list of tuples with the indices and distances.
+        :raises ValueError: If k is less than 1.
+        """
         if k < 1:
-            raise ValueError("num should be >= 1, is now {num}")
+            raise ValueError(f"k should be >= 1, is now {k}")
 
         out: QueryResult = []
+        num_vectors = len(self.vectors)
+        effective_k = min(k, num_vectors)
 
+        # Batch the queries
         for index in range(0, len(vectors), 1024):
             batch = vectors[index : index + 1024]
-            distances = self._dist(batch, self.norm_vectors)
-            if k == 1:
-                sorted_indices = np.argmin(distances, 1, keepdims=True)
-            elif k >= len(self.vectors):
-                # If we want more than we have, just sort everything.
-                sorted_indices = np.stack([np.arange(len(self.vectors))] * len(vectors))
-            else:
-                sorted_indices = np.argpartition(distances, kth=k, axis=1)
-                sorted_indices = sorted_indices[:, :k]
-            for lidx, indices in enumerate(sorted_indices):
-                dists_for_word = distances[lidx, indices]
-                word_index = np.argsort(dists_for_word)
-                i = indices[word_index]
-                d = dists_for_word[word_index]
-                out.append((i, d))
+            distances = self._dist(batch)
 
-        return out
+            # Efficiently get the k smallest distances
+            indices = np.argpartition(distances, kth=effective_k - 1, axis=1)[:, :effective_k]
+            sorted_indices = np.take_along_axis(
+                indices, np.argsort(np.take_along_axis(distances, indices, axis=1)), axis=1
+            )
+            sorted_distances = np.take_along_axis(distances, sorted_indices, axis=1)
 
-    @classmethod
-    def _dist(cls, x: npt.NDArray, y: npt.NDArray) -> npt.NDArray:
-        """Cosine distance function. This assumes y is normalized."""
-        sim = normalize(x).dot(y.T)
+            # Extend the output with tuples of (indices, distances)
+            out.extend(zip(sorted_indices, sorted_distances))
 
-        return 1 - sim
+        return out
 
     def insert(self, vectors: npt.NDArray) -> None:
         """Insert vectors into the vector space."""
         self._vectors = np.vstack([self._vectors, vectors])
+        self._update_precomputed_data()
 
     def delete(self, indices: list[int]) -> None:
         """Deletes specific indices from the vector space."""
         self._vectors = np.delete(self._vectors, indices, axis=0)
+        self._update_precomputed_data()
+
+
+class CosineBasicBackend(BasicBackend):
+    def __init__(self, vectors: npt.NDArray, arguments: BasicArgs) -> None:
+        """Initialize the cosine basic backend."""
+        super().__init__(arguments)
+        self._vectors = vectors
```

Comment:
little optimization: because you know that you are using cosine here you no longer need to store _vectors.